src/core/maprenderer/qgsmaprenderercustompainterjob.cpp

```diff
@@ -343,12 +349,30 @@ void QgsMapRendererCustomPainterJob::doRender()
       mPainter->setOpacity( 1.0 );
     }
 
+    if ( mainElevationMap && job.context()->elevationMap() )
+    {
+      const QgsElevationMap &layerElevationMap = *job.context()->elevationMap();
+      if ( layerElevationMap.isValid() )
+        mainElevationMap->combine( layerElevationMap, mapShadingRenderer.combinedElevationMethod() );
+    }
+
     emit layerRendered( job.layerId );
   }
 
   emit renderingLayersFinished();
   QgsDebugMsgLevel( QStringLiteral( "Done rendering map layers" ), 5 );
 
+  if ( mapShadingRenderer.isActive() &&  mainElevationMap )
+  {
+    QImage image( mainElevationMap->rawElevationImage().size(), QImage::Format_RGB32 );
+    image.fill( Qt::white );
+    mapShadingRenderer.renderShading( *mainElevationMap.get(), image, QgsRenderContext::fromMapSettings( mSettings ) );
+    mPainter->save();
+    mPainter->setCompositionMode( QPainter::CompositionMode_Multiply );
```

Comment:
We should expose the blend mode (and opacity!) as an option for users to control here :+1:

Reply:
hmm, I am no sure t understand what you want. Here, we just shade the rendered map by multiplying it by a grayscale image obtained from the map elevation.

Reply:
Right -- what I'm proposing is that we add the choice of composition (blend) mode to the project elevation rendering page, so that users can select an alternative blending mode if multiply isn't appropriate for their particular map. Ie we add this widget group to the elevation rendering page:

![image](https://user-images.githubusercontent.com/1829991/214185004-441c9a59-7680-4f4b-a044-3f67884c8f09.png)

Opacity control in particular would be quite useful to allow users to "dial down" the effect, and exposing blend mode control is straightforward and will be of use to power users.

Reply:
ah! I see. But in the code you pointed (for `QgMapRenderersCustomPainterjob`), the blending is used because it is the way I found to obtain the shading when the layout is exported (I don't remember well, but I think it is when the layout was exported to PDF, without this blending, the shading was not rendered on the PDF).

For other map renderer, it is the `QgsElevationShadindRenderer` that is applying the shading directly on the rendered image, without using any composition mode. So, it could be interesting to allow the user to choose the blending mode, but not so straightforward to implement.
It could be simpler for opacity, but that need a bit of work.

May be in another PR?

Reply:
> For other map renderer, it is the QgsElevationShadindRenderer that is applying the shading directly on the rendered image, without using any composition mode

Won't that lead to (possibly significant) rendering differences between canvas vs layout exports? I'm ok with the two different rendering paths, but I think the same composition mode would need to be used for both.

Reply:
> Won't that lead to (possibly significant) rendering differences between canvas vs layout exports?

The pipe is different but the method is exactly the same: we multiply by a factor the colors values to darken the map. With `QgMapRenderersCustomPainterjob`, we use an intermediate grayscale image produced by applying the shading on a white image. Then we multiply this image with the map image with `QPainter::CompositionMode_Multiply`
For other map renderer, the factor is applied directly on the map image pixels.
So I don't see any reasons that the rendering is not the same.
In both cases, final pixel values are the same.